FIX: mne.concatenate_raws(raws) wrongly concatenates raws

moritz-gerster opened the pull request:
#### Reference issue
Fixes #11633.

mscheltienne commented on mne/io/base.py:
```diff
@@ -2549,6 +2549,12 @@ def concatenate_raws(raws, preload=None, events_list=None, *,
     events : ndarray of int, shape (n_events, 3)
         The events. Only returned if ``event_list`` is not None.
     """
+    # consistent channel order across raws (not checked by _ensure_infos_match)
+    ch_order = raws[0].ch_names
+    for raw in raws[1:]:
+        raw.reorder_channels(ch_order)
+    # [raw.reorder_channels(raws[0].ch_names) for raw in raws[1:]]
```
This is simple, but it would reorder the `raw[1:]` in place, modifying the user variable outside the scope of this function. I think you need to check one by one if the order is different and then **copy** and re-order only the ones that have a different order.

drammock replied:
proposal: can we make it so that `_ensure_infos_match` *does* check channel order (thereby making `concatenate_raws` fail if they don't) and then provide a new function (`match_channel_orders` or so) that we advise users to run before concatenating?

agramfort replied:
I would mimic the api of concatenation of epochs or the grand average function for evoked

larsoner replied:
@moritz-gerster can you comment here?

moritz-gerster replied:
@larsoner I don't understand, what should I comment? I posted two comments right above your comment.

So far, I implemented a solution that requires further discussion and implemented a test that passes under the current implementation. I am waiting for all of you to agree before I proceed. Currently, @mscheltienne, @drammock, and @agramfort suggested three very different solutions. I am not experienced enough to choose one of them.

agramfort commented:
@moritz-gerster can you add a tiny test using simulated data?

thanks

moritz-gerster commented:
> @moritz-gerster can you add a tiny test using simulated data?

@agramfort done.

mscheltienne commented:
Hello, no we don't see "pending" comments. You probably commented from the tab "Files Changed" and it added your changes as "pending" until you release the review. On the same tab, you need to press on "Review Changes" and "Submit review".

![image](https://user-images.githubusercontent.com/73893616/234087948-2d9d2fcd-b950-44c2-84a8-821cc12c5ad3.png)

drammock commented:
> Can you actually see my comments @larsoner? I am confused because my comments have the label "pending".

Usually that means that on the files changed tab you clicked "start a review" and made several comments but then never clicked "finish your review"

larsoner commented:
+1 for checking if they match and raising an error if they don't, and providing a `match_channel_orders` or so that does the matching. Then we can add options like `copy=True (default) | False` that either returns copies or not without having to add new kwargs to `concatenate_raws`

moritz-gerster commented on mne/io/base.py:
```diff
@@ -2570,6 +2570,15 @@ def concatenate_raws(raws, preload=None, events_list=None, *,
         return raws[0], events
 
 
+def match_channel_orders(raws, copy=True):
+    """Ensure consistent channel order across raws."""
+    raws = deepcopy(raws) if copy else raws
+    ch_order = raws[0].ch_names
+    for raw in raws[1:]:
+        raw.reorder_channels(ch_order)
+    return raws
+
+
```
Is this fine?

drammock replied:
needs a proper docstring (with `Parameters` and `Returns` sections at least). Probably should also add it here:

https://github.com/mne-tools/mne-python/blob/57f5ce300f8e67ffcc0d9a19ab724693105b74b6/mne/__init__.py#L38

and here:

https://github.com/mne-tools/mne-python/blob/57f5ce300f8e67ffcc0d9a19ab724693105b74b6/doc/sensor_space.rst?plain=1#L11

moritz-gerster commented on mne/io/fiff/tests/test_raw_fiff.py:
```diff
@@ -429,6 +430,38 @@ def test_concatenate_raws_bads_order():
+        # Fails now due to wrong order of channels
+        raw_concat = concatenate_raws(raws)
+
+    with pytest.raises(ValueError):
+        # still fails, because raws is copied and not changed in place
+        match_channel_orders(raws, copy=True)
+        raw_concat = concatenate_raws(raws)
+
+    # Now passes because all raws have the same order
+    match_channel_orders(raws, copy=False)
+    raw_concat = concatenate_raws(raws)
+    ch0 = raw_concat.copy().pick(picks=["0"]).get_data()
+    assert np.all(ch0 == 0)
+
```
Here, I decided to test `concatenate_raws` and `match_channel_orders` in a single test.

drammock replied:
fine to test both in one test I think. Please add `match` clauses to the `with pytest.raises()` calls as mentioned in another comment.

moritz-gerster replied:
Adressed in e76a1bd1a0734afc8ebf7b7e8d317f663f5bf0a4

moritz-gerster commented on mne/io/meas_info.py:
```diff
@@ -2906,6 +2906,10 @@ def _ensure_infos_match(info1, info2, name, *, on_mismatch='raise'):
         raise ValueError(f'{name}.info[\'sfreq\'] must match')
     if set(info1['ch_names']) != set(info2['ch_names']):
         raise ValueError(f'{name}.info[\'ch_names\'] must match')
```
I kept this to enable a more precise error message. This way it should be faster to understand for the user whether some channels are missing or whether the channel order is wrong.

moritz-gerster commented:
> +1 for checking if they match and raising an error if they don't, and providing a match_channel_orders or so that does the matching. Then we can add options like copy=True (default) | False that either returns copies or not without having to add new kwargs to concatenate_raws

@larsoner done.

drammock commented on mne/io/fiff/tests/test_raw_fiff.py:
```diff
@@ -389,7 +390,7 @@ def _create_toy_data(n_channels=3, sfreq=250, seed=None):
 
 def test_concatenate_raws_bads_order():
-    """Test concatenation of raw instances."""
+    """Test concatenation of raw instances when the order of bad chs varies."""
```
```suggestion
    """Test concatenation of raws when the order of *bad* channels varies."""
```

drammock commented on mne/io/fiff/tests/test_raw_fiff.py:
```diff
@@ -429,6 +430,38 @@ def test_concatenate_raws_bads_order():
+    # Now passes because all raws have the same order
+    match_channel_orders(raws, copy=False)
+    raw_concat = concatenate_raws(raws)
+    ch0 = raw_concat.copy().pick(picks=["0"]).get_data()
```
```suggestion
    ch0 = raw_concat.get_data(picks=["0"])
```

drammock commented on mne/io/meas_info.py:
```diff
@@ -2906,6 +2906,10 @@ def _ensure_infos_match(info1, info2, name, *, on_mismatch='raise'):
         raise ValueError(f'{name}.info[\'ch_names\'] must match')
+    if info1['ch_names'] != info2['ch_names']:
+        msg = (f'{name}.info[\'ch_names\']: Channel order must match. Use '
+                'mne.match_channel_orders() to sort channels.')
```
does it even work to call it as `mne.match_channel_orders()`? (see my earlier comment about adding it to `mne.__init__.py`)

drammock replied:
no that is not what I meant. What I meant was will it really work for users to do what is recommended in the error message: to say `from mne import match_channel_orders` (or the equivalent `import mne; mne.match_channel_orders`).  I think it will *not* work because the function isn't in `mne/__init__.py` (or at least it wasn't when I reviewed last).

aside: cross-referencing the documentation by using `:func:` is not helpful inside error messages, which are always shown in the terminal / never interpreted as reStructuredText (only docstrings, tutorials, and other webpages need such crossreferences).

moritz-gerster replied:
I've added it in 1a801a638373dbe53aee75b31a1ca96915ea1677

drammock reviewed (commented):
Other than my detailed comments (below) one general comment is that the new test is very thorough!  It might even be overkill; I'll let @larsoner decide if it needs some pruning (our current tests take a really long time to run so we are very cautious about how fast new tests run)

moritz-gerster commented on mne/io/fiff/tests/test_raw_fiff.py:
```diff
@@ -416,7 +417,7 @@ def test_concatenate_raws_bads_order():
     # Type mismatch raises
     epochs1 = make_fixed_length_epochs(raw1)
     with pytest.raises(ValueError):
```
Adressed in e76a1bd1a0734afc8ebf7b7e8d317f663f5bf0a4

drammock replied:
That commit fixed lines 449 and 453 but it didn't do this line.

moritz-gerster replied:
I've added it in 31e5aa24400fcc5ebf883c0030789e1c6bb26a7d

moritz-gerster commented:
> Other than my detailed comments (below) one general comment is that the new test is very thorough! It might even be overkill; I'll let @larsoner decide if it needs some pruning (our current tests take a really long time to run so we are very cautious about how fast new tests run)

thanks @drammock! I addressed all your comments

drammock commented on doc/sensor_space.rst:
```diff
@@ -9,6 +9,7 @@ Sensor Space Data
 
    combine_evoked
    concatenate_raws
+   match_channel_orders
```
this list should remain alphabetized.

drammock commented on mne/io/base.py:
```diff
@@ -2570,6 +2570,27 @@ def concatenate_raws(raws, preload=None, events_list=None, *,
+    Parameters
+    ----------
+    raws : list
+        List of `~mne.io.Raw` instances to order.
```
```suggestion
        List of :class:`~mne.io.Raw` instances to order.
```

drammock commented on mne/io/base.py:
```diff
@@ -2570,6 +2570,27 @@ def concatenate_raws(raws, preload=None, events_list=None, *,
+    ----------
+    raws : list
+        List of `~mne.io.Raw` instances to order.
+    %(copy_df)s
```
this won't work unless the function is decorated with `@fill_doc` or `@verbose` (here use `@fill_doc` because there's no `verbose` parameter)

moritz-gerster commented:
again: thanks @drammock! I addressed all your comments

drammock commented on mne/io/fiff/tests/test_raw_fiff.py:
```diff
@@ -411,22 +411,22 @@ def test_concatenate_raws_bads_order():
     raw2 = raw1.copy()
     raw2.info["bads"] = ["0", "2"]
-    with pytest.raises(ValueError):
+    with pytest.raises(ValueError, match="must match"):
```
```suggestion
    with pytest.raises(ValueError, match="order must match"):
```

drammock commented on mne/io/fiff/tests/test_raw_fiff.py:
```diff
@@ -411,22 +411,22 @@ def test_concatenate_raws_bads_order():
     # Sample rate mismatch
     raw3 = _create_toy_data(sfreq=500)
-    with pytest.raises(ValueError):
+    with pytest.raises(ValueError, match="must match"):
```
```suggestion
    with pytest.raises(ValueError, match="order must match"):
```

moritz-gerster replied:
This does not work @drammock. The error message is 
```Python
ValueError: raws[1].info['bads'] must match
```
since the sample rate does not match.

moritz-gerster replied:
Wait a second, the error message is wrong! I've corrected it in [2670811](https://github.com/mne-tools/mne-python/pull/11640/commits/2670811b9c56339fdfe4b37828a701e6cffc0758)

I can't write 
```Python
with pytest.raises(ValueError, match="raws[1].info['sfreq'] must match"):
    concatenate_raws([raw0, raw3])
```
because then the test fails.

moritz-gerster replied:
The error message is now correct though. I am not sure why matching `raws[1].info['sfreq']` does not work.

larsoner replied:
Because what is matched is a regular expression (regex) the `[]` (character class delimiter in regex) need to be escaped not the quotes. You could also replace them with `.` which is regex for (almost) anything

larsoner replied:
... and finally, a much faster way to iterate is too play with some online regex tool like https://regex101.com/ and see what works

moritz-gerster replied:
Thanks @larsoner, I changed the code! [09f444a](https://github.com/mne-tools/mne-python/pull/11640/commits/09f444a1894ec565dc6f957bdf12f67bef8a503e)

drammock commented on mne/io/fiff/tests/test_raw_fiff.py:
```diff
@@ -411,22 +411,22 @@ def test_concatenate_raws_bads_order():
     # Number of channels mismatch
     raw4 = _create_toy_data(n_channels=4)
-    with pytest.raises(ValueError):
+    with pytest.raises(ValueError, match="must match"):
```
```suggestion
    with pytest.raises(ValueError, match="order must match"):
```

moritz-gerster replied:
This also does not work @drammock. The error message is 
```Python
ValueError: raws[1].info['nchan'] must match
```
since the number of channels does not match.

drammock replied:
oops!  I was incorrectly inferring which error message this was supposed to raise.  I still think the `match` value needs to change though: apparently we have at least 3 errors that are possible to encounter in this context, and all of them contain the string `must match` so we need `match` to be more specific about which one we expect.

drammock reviewed (approved):
Looks good. I've left suggestions to disambiguate the `match` clauses but otherwise I'm happy. CIs are not though, but it looks unrelated (openmeeg / pyvista problems?)

drammock reviewed (approved):
thanks for the more robust `match` arguments. FYI (don't need to do this unless you want to) partial patterns like
`'raws.*info.*nchan.*must match'` or even just `'nchan.*must match'` would have sufficed to disambiguate (with `nchan` / `sfreq` / `bads` substituted accordingly).

drammock commented:
ci failure is this:
```
 _______________________ test_brain_save_movie[pyvistaqt] _______________________
mne/viz/_brain/tests/test_brain.py:530: in test_brain_save_movie
    brain.save_movie(filename, time_dilation=1, tmin=1, tmax=1.1,
mne/viz/_brain/_brain.py:3524: in save_movie
    func(filename, time_dilation, tmin, tmax,
mne/viz/_brain/_brain.py:3419: in _save_movie
    import imageio
E   ModuleNotFoundError: No module named 'imageio'
```
??? I looked at their recent commit history, nothing jumps out as an obvious culprit.

larsoner commented:
Thanks @moritz-gerster !
